Sming/Libraries/OtaUpgrade/BasicOtaUpgradeStream.cpp

```diff
@@ -0,0 +1,251 @@
+		}
+	}
+
+	for(auto offset : OTAUpgrade_AppFlashRegionOffsets) {
+		limitSize(offset);
+	}
+}
+
+BasicOtaUpgradeStream::BasicOtaUpgradeStream()
+{
+	setupChunk(StateHeader, fileHeader);
+}
+
+bool BasicOtaUpgradeStream::consume(const uint8_t*& data, size_t& size)
+{
+	size_t chunkSize = std::min(size, remainingBytes);
+	if(state != StateVerifyRoms) {
+		verifier.update(data, chunkSize);
+	}
+
+	if(destinationPtr != nullptr) {
+		memcpy(destinationPtr, data, chunkSize);
+		destinationPtr += chunkSize;
+	}
+	remainingBytes -= chunkSize;
+	data += chunkSize;
+	size -= chunkSize;
+	if(remainingBytes == 0) {
+		destinationPtr = nullptr;
+		return true;
+	} else {
+		return false;
+	}
+}
+
+void BasicOtaUpgradeStream::setError(ErrorCode code)
+{
+	assert(code != NoError);
+	debug_e("Error: %s", errorToString(code).c_str());
+	errorCode = code;
+	state = StateError;
+}
+
+void BasicOtaUpgradeStream::nextRom()
+{
+	if(romIndex < fileHeader.romCount) {
+		++romIndex;
+		setupChunk(StateRomHeader, romHeader);
+	} else {
+		setupChunk(StateVerifyRoms, verificationData);
+	}
+}
+
+void BasicOtaUpgradeStream::processRomHeader()
+{
+	bool addressMatch = (slot.address & 0xFFFFF) == (romHeader.address & 0xFFFFF);
+	if(!slot.updated && addressMatch) {
+		if(romHeader.size <= slot.size) {
+			debug_i("Update slot %u [0x%08X..0x%08X)", slot.index, slot.address, slot.address + romHeader.size);
+			rbootWriteStatus = rboot_write_init(slot.address);
+			setupChunk(StateWriteRom, romHeader.size);
+		} else {
+			setError(RomTooLargeError);
+		}
+		return;
+	}
+
+	debug_i("Skip ROM image for [0x%08X..0x%08X)", romHeader.address, romHeader.address + romHeader.size);
+	setupChunk(StateSkipRom, romHeader.size);
+}
+
+void BasicOtaUpgradeStream::verifyRoms()
+{
+	state = StateRomsComplete;
+
+	debug_d("Signature/Checksum: %s", makeHexString(verificationData, sizeof(verificationData), ' ').c_str());
+
+	if(!verifier.verify(verificationData)) {
+		if(slot.updated) {
+			// Destroy start sector of updated ROM to avoid accidentally booting an unsanctioned firmware
+			flashmem_erase_sector(slot.address / SECTOR_SIZE);
+		}
+		setError(VerificationError);
+		return;
+	}
+
+	// In a future extension, there might be OTA files without ROM images (SPIFFS-only update, etc.)
+	if(fileHeader.romCount == 0) {
+		return;
+	}
+
+	debug_i("ROM update complete");
+	if(!slot.updated) {
+		setError(NoRomFoundError);
+		return;
+	}
+
+	if(rboot_set_current_rom(slot.index)) {
+		debug_i("ROM %u activated", slot.index);
+	} else {
+		setError(RomActivationError);
+	}
+}
+
+size_t BasicOtaUpgradeStream::write(const uint8_t* data, size_t size)
+{
+	const size_t origSize = size;
+
+	while(!hasError() && (size > 0)) {
+		switch(state) {
+		case StateHeader:
+			if(consume(data, size)) {
+				if(fileHeader.magic == expectedHeaderMagic) {
+#ifndef ENABLE_OTA_DOWNGRADE
+					uint64_t buildTimestampFirmware;
+					memcpy_P(&buildTimestampFirmware, &OTA_BuildTimestamp, sizeof(OTA_BuildTimestamp));
```

Comment:
`auto buildTimestampFirmware = FSTR::readValue(&OTA_BuildTimestamp);`

Reply:
That's what I was looking for :+1: 
I should take more time to familiarize myself with the FlashString library.

BTW: Do you think it would be worth to always add a build timestamp as part of the core build process (as part of the final link step)? I can think of more uses for it than just the OTAUpgrade library. 
Also, the current solution triggers a relink every time 'make' is invoked, which is currently not a big issue, since it is the default behaviour of the build system, but maybe this will change in the future.

Reply:
Short answer: Yes.

I've done this previously just using __DATE__ and __TIME__ but a programmatic version would certainly be helpful. Might even consider implementing as a set of structured metadata so it can be dumped onto a web page as well as being easily parsed by the application. Could include a list of Components with their name/version, etc. I suspect this could be done very easily using a custom FSTR object.

Reply:
That's similar to what I had in mind. However, since all that data might become quite large, it is probably better to split it into multiple flash objects, probably with a nice frontend API, to allow the linker to drop unreferenced information.